File: app/src/main/java/com/kumi/Bahamut/Pages/MailBoxPage.java
package com.kumi.Bahamut.Pages;

import android.view.View;
import android.view.ViewGroup;
import android.widget.AbsListView;
import android.widget.AdapterView;
import android.widget.Button;
import android.widget.ListAdapter;
import android.widget.ListView;
import com.kumi.ASFramework.Dialog.ASAlertDialog;
import com.kumi.ASFramework.Dialog.ASAlertDialogListener;
import com.kumi.ASFramework.PageController.ASViewController;
import com.kumi.ASFramework.Thread.ASRunner;
import com.kumi.ASFramework.UI.ASToast;
import com.kumi.Bahamut.Command.BahamutCommandDeleteArticle;
import com.kumi.Bahamut.Command.BahamutCommandGoodArticle;
import com.kumi.Bahamut.Command.BahamutCommandSearchArticle;
import com.kumi.Bahamut.Command.BahamutCommandSendMail;
import com.kumi.Bahamut.Command.TelnetCommand;
import com.kumi.Bahamut.Dialogs.Dialog_SearchArticle_Listener;
import com.kumi.Bahamut.Dialogs.Dialog_SelectArticle;
import com.kumi.Bahamut.Dialogs.Dialog_SelectArticle_Listener;
import com.kumi.Bahamut.ListPage.TelnetListPage;
import com.kumi.Bahamut.ListPage.TelnetListPageBlock;
import com.kumi.Bahamut.ListPage.TelnetListPageItem;
import com.kumi.Bahamut.Pages.Model.MailBoxPageBlock;
import com.kumi.Bahamut.Pages.Model.MailBoxPageHandler;
import com.kumi.Bahamut.Pages.Model.MailBoxPageItem;
import com.kumi.Telnet.Logic.ItemUtils;
import com.kumi.Telnet.TelnetOutputBuilder;
import com.kumi.TelnetUI.TelnetHeaderItemView;
import java.util.Vector;

public class MailBoxPage extends TelnetListPage implements ListAdapter, Dialog_SearchArticle_Listener, Dialog_SelectArticle_Listener, SendMailPage_Listener, View.OnClickListener, View.OnLongClickListener {
  private Button _back_button = null;
  
  private TelnetHeaderItemView _header_view = null;
  
  private View _list_empty_view = null;
  
  private Button _page_down_button = null;
  
  private Button _page_up_button = null;
  
  private void onDeleteArticle(final int itemIndex) {
    ASAlertDialog.createDialog().setTitle("刪除").setMessage("是否確定要刪除此信件?").addButton("取消").addButton("刪除").setListener(new ASAlertDialogListener() {
          final MailBoxPage this$0;
          
          final int val$itemIndex;
          
          public void onAlertDialogDismissWithButtonIndex(ASAlertDialog param1ASAlertDialog, int param1Int) {
            switch (param1Int) {
              default:
                return;
              case 1:
                break;
            } 
            BahamutCommandDeleteArticle bahamutCommandDeleteArticle = new BahamutCommandDeleteArticle(itemIndex);
            MailBoxPage.this.pushCommand((TelnetCommand)bahamutCommandDeleteArticle);
          }
        }).scheduleDismissOnPageDisappear((ASViewController)this).show();
  }
  
  private void onPostButtonClicked() {
    SendMailPage sendMailPage = new SendMailPage();
    sendMailPage.setListener(this);
    getNavigationController().pushViewController((ASViewController)sendMailPage);
  }
  
  private void showSelectArticleDialog() {
    Dialog_SelectArticle dialog_SelectArticle = new Dialog_SelectArticle();
    dialog_SelectArticle.setListener(this);
    dialog_SelectArticle.show();
  }
  
  public void clear() {
    super.clear();
  }
  
  public String getListName() {
    return "[MailBox]";
  }
  
  public int getPageLayout() {
    return 2131361844;
  }
  
  public int getPageType() {
    return 9;
  }
  
  public View getView(int paramInt, View paramView, ViewGroup paramViewGroup) {
    MailBoxPage_ItemView mailBoxPage_ItemView2;
    int i = paramInt + 1;
    int j = ItemUtils.getBlock(i);
    MailBoxPageItem mailBoxPageItem = (MailBoxPageItem)getItem(paramInt);
    if (mailBoxPageItem == null && getCurrentBlock() != j && !isLoadingBlock(i))
      loadBoardBlock(j); 
    View view = paramView;
    if (paramView == null) {
      mailBoxPage_ItemView2 = new MailBoxPage_ItemView(getContext());
      mailBoxPage_ItemView2.setLayoutParams((ViewGroup.LayoutParams)new AbsListView.LayoutParams(-1, -2));
    } 
    MailBoxPage_ItemView mailBoxPage_ItemView1 = mailBoxPage_ItemView2;
    mailBoxPage_ItemView1.setItem(mailBoxPageItem);
    mailBoxPage_ItemView1.setIndex(paramInt + 1);
    return (View)mailBoxPage_ItemView2;
  }
  
  public void goodArticle(final int articleIndex) {
    ASAlertDialog.createDialog().setTitle("推薦").setMessage("是否要推薦此文章?").addButton("取消").addButton("推薦").setListener(new ASAlertDialogListener() {
          final MailBoxPage this$0;
          
          final int val$articleIndex;
          
          public void onAlertDialogDismissWithButtonIndex(ASAlertDialog param1ASAlertDialog, int param1Int) {
            switch (param1Int) {
              default:
                return;
              case 1:
                break;
            } 
            BahamutCommandGoodArticle bahamutCommandGoodArticle = new BahamutCommandGoodArticle(articleIndex);
            MailBoxPage.this.pushCommand((TelnetCommand)bahamutCommandGoodArticle);
          }
        }).scheduleDismissOnPageDisappear((ASViewController)this).show();
  }
  
  public void goodLoadingArticle() {
    goodArticle(getLoadingItemNumber());
  }
  
  public boolean isAutoLoadEnable() {
    return false;
  }
  
  public void loadFirstArticle() {
    if (getLoadingItemNumber() == 1) {
      ASToast.showShortToast("已讀至列首");
      return;
    } 
    loadItemAtNumber(1);
  }
  
  public void loadLastestArticle() {
    if (getLoadingItemNumber() == getItemSize()) {
      ASToast.showShortToast("已讀至列尾");
      return;
    } 
    loadItemAtNumber(getItemSize());
  }
  
  public void loadNextArticle() {
    int i = getLoadingItemNumber() + 1;
    if (i > getItemSize()) {
      ASToast.showShortToast("已讀至列尾");
      return;
    } 
    loadItemAtNumber(i);
  }
  
  public TelnetListPageBlock loadPage() {
    return (TelnetListPageBlock)MailBoxPageHandler.getInstance().load();
  }
  
  public void loadPreviousArticle() {
    int i = getLoadingItemNumber() - 1;
    if (i < 1) {
      ASToast.showShortToast("已讀至列首");
      return;
    } 
    loadItemAtNumber(i);
  }
  
  protected boolean onBackPressed() {
    clear();
    getNavigationController().popViewController();
    TelnetOutputBuilder.create().pushKey(256).pushKey(256).sendToServerInBackground(1);
    return true;
  }
  
  public void onClick(View paramView) {
    switch (paramView.getId()) {
      default:
        return;
      case 2131230881:
        onPostButtonClicked();
      case 2131230888:
        moveToFirstPosition();
      case 2131230887:
        setManualLoadPage();
        moveToLastPosition();
      case 2131230889:
        break;
    } 
    showSelectArticleDialog();
  }
  
  public boolean onItemLongClick(AdapterView<?> paramAdapterView, View paramView, int paramInt, long paramLong) {
    onDeleteArticle(paramInt + 1);
    return true;
  }
  
  public boolean onLongClick(View paramView) {
    boolean bool2 = false;
    boolean bool1 = bool2;
    switch (paramView.getId()) {
      default:
        bool1 = bool2;
      case 2131230881:
        return bool1;
      case 2131230888:
        bool1 = true;
      case 2131230887:
        break;
    } 
    bool1 = true;
  }
  
  public void onPageDidDisappear() {
    this._back_button = null;
    this._page_up_button = null;
    this._page_down_button = null;
    this._header_view = null;
    this._list_empty_view = null;
    super.onPageDidDisappear();
  }
  
  public void onPageDidLoad() {
    super.onPageDidLoad();
    ListView listView = (ListView)findViewById(2131230878);
    this._list_empty_view = findViewById(2131230877);
    listView.setEmptyView(this._list_empty_view);
    setListView(listView);
    this._back_button = (Button)findViewById(2131230881);
    this._back_button.setOnClickListener(this);
    this._back_button.setOnLongClickListener(this);
    this._page_up_button = (Button)findViewById(2131230888);
    this._page_up_button.setOnClickListener(this);
    this._page_up_button.setOnLongClickListener(this);
    this._page_down_button = (Button)findViewById(2131230887);
    this._page_down_button.setOnClickListener(this);
    this._page_down_button.setOnLongClickListener(this);
    ((Button)findViewById(2131230889)).setOnClickListener(this);
    this._header_view = (TelnetHeaderItemView)findViewById(2131230880);
  }
  
  public void onPageRefresh() {
    // Byte code:
    //   0: aload_0
    //   1: monitorenter
    //   2: aload_0
    //   3: invokespecial onPageRefresh : ()V
    //   6: aload_0
    //   7: getfield _header_view : Lcom/kumi/TelnetUI/TelnetHeaderItemView;
    //   10: astore_1
    //   11: new java/lang/StringBuilder
    //   14: astore_2
    //   15: aload_2
    //   16: invokespecial <init> : ()V
    //   19: aload_1
    //   20: ldc_w '我的信箱'
    //   23: aload_2
    //   24: ldc_w '您有 '
    //   27: invokevirtual append : (Ljava/lang/String;)Ljava/lang/StringBuilder;
    //   30: aload_0
    //   31: invokevirtual getItemSize : ()I
    //   34: invokevirtual append : (I)Ljava/lang/StringBuilder;
    //   37: ldc_w ' 封信在信箱內'
    //   40: invokevirtual append : (Ljava/lang/String;)Ljava/lang/StringBuilder;
    //   43: invokevirtual toString : ()Ljava/lang/String;
    //   46: ldc_w ''
    //   49: invokevirtual setData : (Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;)V
    //   52: aload_0
    //   53: monitorexit
    //   54: return
    //   55: astore_1
    //   56: aload_0
    //   57: monitorexit
    //   58: aload_1
    //   59: athrow
    // Exception table:
    //   from	to	target	type
    //   2	52	55	finally
  }
  
  public boolean onReceivedGestureRight() {
    onBackPressed();
    ASToast.showShortToast("返回");
    return true;
  }
  
  protected boolean onSearchButtonClicked() {
    showSelectArticleDialog();
    return true;
  }
  
  public void onSearchDialogSearchButtonClickedWithValues(Vector<String> paramVector) {
    String str1;
    String str3 = paramVector.get(0);
    String str2 = paramVector.get(1);
    if (paramVector.get(2) == "YES") {
      str1 = "y";
    } else {
      str1 = "n";
    } 
    pushCommand((TelnetCommand)new BahamutCommandSearchArticle(str3, str2, str1, paramVector.get(3)));
  }
  
  public void onSelectDialogDismissWIthIndex(String paramString) {
    int i = -1;
    try {
      int j = Integer.parseInt(paramString);
      i = j - 1;
    } catch (Exception exception) {
      exception.printStackTrace();
    } 
    if (i >= 0)
      setListViewSelection(i); 
  }
  
  public void onSendMailDialogSendButtonClicked(SendMailPage paramSendMailPage, String paramString1, String paramString2, String paramString3) {
    pushCommand((TelnetCommand)new BahamutCommandSendMail(paramString1, paramString2, paramString3));
  }
  
  public void recoverPost() {
    (new ASRunner() {
        final MailBoxPage this$0;
        
        public void run() {}
      }).runInMainThread();
  }
  
  public void recycleBlock(TelnetListPageBlock paramTelnetListPageBlock) {
    MailBoxPageBlock.recycle((MailBoxPageBlock)paramTelnetListPageBlock);
  }
  
  public void recycleItem(TelnetListPageItem paramTelnetListPageItem) {
    MailBoxPageItem.recycle((MailBoxPageItem)paramTelnetListPageItem);
  }
}


/* Location:              C:\Users\kodak\Downloads\反編譯\dex-tools-v2.4\classes-dex2jar.jar!\com\kumi\Bahamut\Pages\MailBoxPage.class
 * Java compiler version: 6 (50.0)
 * JD-Core Version:       1.1.3
 */

File: app/src/main/java/com/kota/Bahamut/pages/mailPage/MailBoxPage.kt
package com.kota.Bahamut.pages.mailPage

import android.content.Context
import android.database.DataSetObserver
import android.util.Log
import android.view.View
import android.widget.ListAdapter
import androidx.compose.foundation.ExperimentalFoundationApi
import androidx.compose.foundation.background
import androidx.compose.foundation.combinedClickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.IntrinsicSize
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.ComposeView
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import com.kota.Bahamut.BahamutPage
import com.kota.Bahamut.R
import com.kota.Bahamut.command.BahamutCommandDeleteArticle
import com.kota.Bahamut.command.BahamutCommandSearchArticle
import com.kota.Bahamut.command.BahamutCommandSendMail
import com.kota.Bahamut.command.TelnetCommand
import com.kota.Bahamut.dialogs.DialogSearchArticleListener
import com.kota.Bahamut.dialogs.DialogSelectArticle
import com.kota.Bahamut.dialogs.DialogSelectArticleListener
import com.kota.Bahamut.listPage.TelnetListPage
import com.kota.Bahamut.listPage.TelnetListPageBlock
import com.kota.Bahamut.listPage.TelnetListPageItem
import com.kota.Bahamut.pages.model.MailBoxPageBlock
import com.kota.Bahamut.pages.model.MailBoxPageHandler
import com.kota.Bahamut.pages.model.MailBoxPageItem
import com.kota.Bahamut.service.CommonFunctions.getContextString
import com.kota.Bahamut.ui.components.BBSToolbar
import com.kota.Bahamut.ui.components.BBSToolbarDivider
import com.kota.Bahamut.ui.components.BBSTopBar
import com.kota.Bahamut.ui.components.BahaButton
import com.kota.Bahamut.ui.components.BahaText
import com.kota.Bahamut.ui.components.BahaTextSize
import com.kota.Bahamut.ui.components.ButtonType
import com.kota.Bahamut.ui.components.RightArrow
import com.kota.Bahamut.ui.dialogs.BahaGlobalDialogHost
import com.kota.Bahamut.ui.theme.AppTheme
import com.kota.Bahamut.ui.theme.setBahamutContent
import com.kota.asFramework.dialog.ASAlertDialog
import com.kota.asFramework.ui.ASToast.showShortToast
import com.kota.telnet.TelnetOutputBuilder.Companion.create
import com.kota.telnet.logic.ItemUtils
import com.kota.telnet.reference.TelnetKeyboard
import kotlinx.coroutines.launch
import java.util.Vector
import kotlin.math.max

@OptIn(ExperimentalFoundationApi::class)
class MailBoxPage : TelnetListPage(), ListAdapter, DialogSearchArticleListener,
    DialogSelectArticleListener, SendMailPageListener {

    // Compose state
    var headerTitleState by mutableStateOf("我的信箱")
    var headerSubtitleState by mutableStateOf("")

    override val pageType: Int
        get() = BahamutPage.BAHAMUT_MAIL_BOX

    override val pageLayout: Int
        get() = 0

    override fun createPageView(context: Context): View {
        return ComposeView(context).apply {
            setBahamutContent {
                MailBoxPageContent()
                BahaGlobalDialogHost()
            }
        }
    }

    override fun loadPage(): TelnetListPageBlock? {
        return MailBoxPageHandler.instance.load()
    }

    @Synchronized
    override fun onPageRefresh() {
        super.onPageRefresh()
        val myListCount = getItemSize()
        headerTitleState = "我的信箱"
        headerSubtitleState = "您有 $myListCount 封信在信箱內"
    }

    override fun onBackPressed(): Boolean {
        clear()
        navigationController.popViewController()
        create().pushKey(TelnetKeyboard.LEFT_ARROW).pushKey(TelnetKeyboard.LEFT_ARROW)
            .sendToServerInBackground(1)
        return true
    }

    override fun onListViewItemLongClicked(itemView: View?, index: Int): Boolean {
        if (isItemCanLoadAtIndex(index)) {
            onDeleteArticle(index + 1)
            return true
        }
        return false
    }

    fun showSelectArticleDialog() {
        val dialog = DialogSelectArticle()
        dialog.setListener(this)
        dialog.show()
    }

    override fun onSearchDialogSearchButtonClickedWithValues(vector: Vector<String>) {
        pushCommand(
            BahamutCommandSearchArticle(
                vector[0]!!,
                vector[1],
                if (vector[2] == "YES") "y" else "n",
                vector[3]
            )
        )
    }

    override fun onSelectDialogDismissWIthIndex(str: String) {
        var itemIndex = -1
        try {
            itemIndex = str.toInt() - 1
        } catch (e: Exception) {
            Log.e(javaClass.simpleName, (if (e.message != null) e.message else "")!!)
        }
        if (itemIndex >= 0) {
            setListViewSelection(itemIndex)
        }
    }

    override fun onSendMailDialogSendButtonClicked(
        sendMailPage: SendMailPage,
        receiver: String,
        title: String,
        content: String
    ) {
        pushCommand(BahamutCommandSendMail(receiver, title, content))
    }

    override fun isItemCanLoadAtIndex(index: Int): Boolean {
        val mailBoxPageItem = getItem(index) as MailBoxPageItem?
        if (mailBoxPageItem == null || mailBoxPageItem.isDeleted) {
            showShortToast("此信件已被刪除")
            return false
        }
        return true
    }

    fun onDeleteArticle(itemIndex: Int) {
        ASAlertDialog.createDialog()
            .setTitle(getContextString(R.string.delete))
            .setMessage(getContextString(R.string.del_this_mail))
            .addButton(getContextString(R.string.cancel))
            .addButton(getContextString(R.string.delete))
            .setListener { _: ASAlertDialog?, index: Int ->
                if (index == 1) {
                    val mailBoxPageItem = getItem(itemIndex - 1) as MailBoxPageItem?
                    mailBoxPageItem?.isDeleted = true
                    val command: TelnetCommand = BahamutCommandDeleteArticle(itemIndex)
                    this@MailBoxPage.pushCommand(command)
                }
            }.scheduleDismissOnPageDisappear(this).show()
    }

    override fun onReceivedGestureRight(): Boolean {
        onBackPressed()
        showShortToast(getContextString(R.string._back))
        return true
    }

    fun onPostButtonClicked() {
        val sendMailPage = SendMailPage()
        sendMailPage.setListener(this)
        navigationController.pushViewController(sendMailPage)
    }

    fun loadPreviousArticle() {
        val targetNumber = loadingItemNumber - 1
        if (targetNumber < 1) {
            showShortToast(getContextString(R.string.already_to_top))
        } else {
            loadItemAtNumber(targetNumber)
        }
    }

    fun loadNextArticle() {
        val targetIndex = loadingItemNumber + 1
        if (targetIndex > getItemSize()) {
            showShortToast(getContextString(R.string.already_to_bottom))
        } else {
            loadItemAtNumber(targetIndex)
        }
    }

    override val isAutoLoadEnable: Boolean
        get() = false

    override var listName: String = ""
        get() = "[MailBox]"

    override fun recycleBlock(telnetListPageBlock: TelnetListPageBlock) {
        MailBoxPageBlock.recycle(telnetListPageBlock as MailBoxPageBlock)
    }

    override fun recycleItem(telnetListPageItem: TelnetListPageItem) {
        MailBoxPageItem.recycle(telnetListPageItem as MailBoxPageItem)
    }

    fun recoverPost() {}
    fun finishPost() {}

    override fun onSearchDialogCancelButtonClicked() {}

    // ---------------------------------------------------------------
    // Compose UI
    // ---------------------------------------------------------------

    @Composable
    fun MailBoxPageContent() {
        val colors = AppTheme.colors
        val coroutineScope = rememberCoroutineScope()
        val listState = rememberLazyListState()

        var dataVersion by remember { mutableIntStateOf(0) }
        DisposableEffect(Unit) {
            val observer = object : DataSetObserver() {
                override fun onChanged() { dataVersion++ }
                override fun onInvalidated() { dataVersion++ }
            }
            registerDataSetObserver(observer)
            onDispose { unregisterDataSetObserver(observer) }
        }

        val currentCount = if (dataVersion >= 0) count else 0

        Column(
            modifier = Modifier
                .fillMaxSize()
                .background(colors.pageBackground)
        ) {
            // 頂部導覽列 (BBS 經典深海藍底，雙行文字資訊)
            BBSTopBar(
                title = headerTitleState.ifEmpty { "我的信箱" },
                subtitle = headerSubtitleState
            )

            // 郵件列表
            Box(
                modifier = Modifier
                    .weight(1f)
                    .fillMaxWidth()
            ) {
                if (currentCount == 0) {
                    Box(
                        modifier = Modifier.fillMaxSize(),
                        contentAlignment = Alignment.Center
                    ) {
                        BahaText(
                            text = stringResource(R.string.mainbox_no_mail),
                            color = colors.textSecondary,
                            fontSize = BahaTextSize.BODY
                        )
                    }
                } else {
                    LazyColumn(state = listState, modifier = Modifier.fillMaxSize()) {
                        items(
                            count = currentCount,
                            key = { index -> index }
                        ) { index ->
                            @Suppress("UNUSED_VARIABLE")
                            val version = dataVersion
                            val itemIndex = index + 1
                            val block = ItemUtils.getBlock(itemIndex)
                            val item = getItem(index) as? MailBoxPageItem
                            if (item == null && currentBlock != block && !isLoadingBlock(itemIndex)) {
                                loadBoardBlock(block)
                            }
                            MailBoxRowItem(
                                item = item,
                                itemIndex = itemIndex,
                                onClick = {
                                    coroutineScope.launch {
                                        loadItemAtIndex(index)
                                    }
                                },
                                onLongClick = {
                                    coroutineScope.launch {
                                        onListViewItemLongClicked(null, index)
                                    }
                                }
                            )
                        }
                    }
                }
            }

            // 底部工具列 (BBS 50dp 全寬操作列)
            BBSToolbar {
                BahaButton(
                    text = stringResource(R.string.write_mail),
                    type = ButtonType.NORMAL,
                    fontSize = BahaTextSize.TITLE,
                    modifier = Modifier
                        .weight(1f)
                        .fillMaxHeight(),
                    onClick = { onPostButtonClicked() }
                )
                BBSToolbarDivider()
                BahaButton(
                    text = stringResource(R.string.search),
                    type = ButtonType.NORMAL,
                    fontSize = BahaTextSize.TITLE,
                    modifier = Modifier
                        .weight(1f)
                        .fillMaxHeight(),
                    onClick = { showSelectArticleDialog() }
                )
                BBSToolbarDivider()
                BahaButton(
                    text = stringResource(R.string.first_page),
                    type = ButtonType.NORMAL,
                    fontSize = BahaTextSize.TITLE,
                    modifier = Modifier
                        .weight(1f)
                        .fillMaxHeight(),
                    onClick = {
                        moveToFirstPosition()
                        coroutineScope.launch { listState.scrollToItem(0) }
                        showShortToast(getContextString(R.string.already_to_top))
                    }
                )
                BBSToolbarDivider()
                BahaButton(
                    text = stringResource(R.string.last_page),
                    type = ButtonType.NORMAL,
                    fontSize = BahaTextSize.TITLE,
                    modifier = Modifier
                        .weight(1f)
                        .fillMaxHeight(),
                    onClick = {
                        setManualLoadPage()
                        moveToLastPosition()
                        coroutineScope.launch { listState.scrollToItem(max(0, currentCount - 1)) }
                        showShortToast(getContextString(R.string.already_to_bottom))
                    }
                )
            }
        }
    }

    @Composable
    private fun MailBoxRowItem(
        item: MailBoxPageItem?,
        itemIndex: Int,
        onClick: () -> Unit,
        onLongClick: () -> Unit
    ) {
        val colors = AppTheme.colors
        val isRead = item?.isRead ?: false
        val textColor = if (isRead) colors.bbsBoardNormalRead else colors.bbsBoardNormal
        val statusText = if (isRead) "◇" else "◆"
        val isMarked = item?.isMarked == true
        val isReply = item?.isReply == true

        Column(
            modifier = Modifier
                .fillMaxWidth()
                .background(colors.pageBackground)
        ) {
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .height(IntrinsicSize.Min),
                verticalAlignment = Alignment.CenterVertically
            ) {
                // 左側信件資訊區塊
                Column(
                    modifier = Modifier
                        .weight(1f)
                        .combinedClickable(
                            onClick = onClick,
                            onLongClick = onLongClick
                        )
                        .padding(horizontal = 10.dp, vertical = 10.dp)
                ) {
                    // 第一列：狀態圖示與信件標題
                    Row(
                        modifier = Modifier.fillMaxWidth(),
                        verticalAlignment = Alignment.CenterVertically
                    ) {
                        BahaText(
                            text = statusText,
                            color = colors.bbsMailStatus,
                            fontSize = BahaTextSize.TITLE,
                            modifier = Modifier.padding(end = 6.dp)
                        )
                        BahaText(
                            text = item?.title ?: stringResource(R.string.loading_),
                            color = textColor,
                            fontSize = BahaTextSize.TITLE,
                            maxLines = 1,
                            overflow = TextOverflow.Ellipsis
                        )
                    }

                    Spacer(modifier = Modifier.height(2.dp))

                    // 第二列：編號、標記(M)、回信(R)、日期、作者
                    Row(
                        modifier = Modifier.fillMaxWidth(),
                        verticalAlignment = Alignment.CenterVertically
                    ) {
                        // 編號 (5 碼)
                        BahaText(
                            text = String.format("%05d", itemIndex),
                            color = colors.bbsMailNumber,
                            fontSize = BahaTextSize.BODY
                        )

                        // 標記 M
                        if (isMarked) {
                            Spacer(modifier = Modifier.width(8.dp))
                            BahaText(
                                text = "M",
                                color = colors.bbsMailMark,
                                fontSize = BahaTextSize.BODY
                            )
                        } else {
                            Spacer(modifier = Modifier.width(42.dp))
                        }

                        // 回信 R
                        if (isReply) {
                            Spacer(modifier = Modifier.width(8.dp))
                            BahaText(
                                text = "R",
                                color = colors.bbsMailReply,
                                fontSize = BahaTextSize.BODY
                            )
                        } else {
                            Spacer(modifier = Modifier.width(42.dp))
                        }

                        // 日期
                        Spacer(modifier = Modifier.width(12.dp))
                        BahaText(
                            text = item?.date ?: "",
                            color = colors.bbsMailDate,
                            fontSize = BahaTextSize.BODY
                        )

                        Spacer(modifier = Modifier.weight(1f))

                        // 作者 (靠右對齊)
                        BahaText(
                            text = item?.author ?: "",
                            color = colors.bbsMailAuthor,
                            fontSize = BahaTextSize.BODY,
                            maxLines = 1,
                            overflow = TextOverflow.Ellipsis
                        )
                    }
                }

                // 右側箭頭按鈕區塊
                RightArrow { onClick() }
            }

            // 底部分隔線
            Box(
                modifier = Modifier
                    .fillMaxWidth()
                    .height(1.dp)
                    .background(colors.divider)
            )
        }
    }
}
